src/plots/mapbox/mapbox.js

```diff
@@ -0,0 +1,377 @@
+/**
+* Copyright 2012-2016, Plotly, Inc.
+* All rights reserved.
+*
+* This source code is licensed under the MIT license found in the
+* LICENSE file in the root directory of this source tree.
+*/
+
+
+'use strict';
+
+var mapboxgl = require('mapbox-gl');
+
+var Fx = require('../cartesian/graph_interact');
+var constants = require('./constants');
+var layoutAttributes = require('./layout_attributes');
+var createMapboxLayer = require('./layers');
+
+
+function Mapbox(opts) {
+    this.id = opts.id;
+    this.gd = opts.gd;
+    this.container = opts.container;
+    this.isStatic = opts.staticPlot;
+
+    var fullLayout = opts.fullLayout;
+
+    // unique id for this Mapbox instance
+    this.uid = fullLayout._uid + '-' + this.id;
+
+    // full mapbox options (N.B. needs to be updated on every updates)
+    this.opts = fullLayout[this.id];
+
+    // create framework on instantiation for a smoother first plot call
+    this.div = null;
+    this.xaxis = null;
+    this.yaxis = null;
+    this.createFramework(fullLayout);
+
+    // state variables used to infer how and what to update
+    this.map = null;
+    this.styleUrl = null;
+    this.traceHash = {};
+    this.layerList = [];
+}
+
+var proto = Mapbox.prototype;
+
+module.exports = function createMapbox(opts) {
+    var mapbox = new Mapbox(opts);
+
+    return mapbox;
+};
+
+proto.plot = function(calcData, fullLayout, promises) {
+    var self = this;
+
+    // feed in new mapbox options
+    self.opts = fullLayout[this.id];
+
+    var promise;
+
+    if(!self.map) {
+        promise = new Promise(function(resolve, reject) {
+            self.createMap(calcData, fullLayout, resolve, reject);
+        });
+    }
+    else {
+        promise = new Promise(function(resolve, reject) {
+            self.updateMap(calcData, fullLayout, resolve, reject);
+        });
+    }
+
+    promises.push(promise);
+};
+
+proto.createMap = function(calcData, fullLayout, resolve, reject) {
+    var self = this,
+        gd = self.gd,
+        opts = self.opts;
+
+    // mapbox doesn't have a way to get the current style URL; do it ourselves
+    var styleUrl = self.styleUrl = convertStyleUrl(opts.style);
+
+    var map = self.map = new mapboxgl.Map({
+        container: self.div,
+
+        style: styleUrl,
+        center: convertCenter(opts.center),
+        zoom: opts.zoom,
+        bearing: opts.bearing,
+        pitch: opts.pitch,
+
+        interactive: !self.isStatic,
+        preserveDrawingBuffer: self.isStatic
+    });
+
+    self.rejectOnError(reject);
+
+    map.once('load', function() {
+        self.updateData(calcData);
+        self.updateLayout(fullLayout);
+
+        self.resolveOnRender(resolve);
+    });
+
+    // keep track of pan / zoom in user layout
+    map.on('move', function() {
+        var center = map.getCenter();
+        opts._input.center = opts.center = { lon: center.lng, lat: center.lat };
+        opts._input.zoom = opts.zoom = map.getZoom();
```

Comment:
`opts._input` here is e.g. `gd.layout.mapbox`,

On every mouse move (drag, pan, zoom ...), we update the input container to mock what `Plotly.relayout` would do.